diff --git a/policy/modules/services/apache.te b/policy/modules/services/apache.te
index b595396..3bcbd71 100644
--- a/policy/modules/services/apache.te
+++ b/policy/modules/services/apache.te
@@ -521,6 +521,7 @@ fs_getattr_all_fs(httpd_t)
 fs_search_auto_mountpoints(httpd_t)
 fs_read_iso9660_files(httpd_t)
 fs_read_anon_inodefs_files(httpd_t)
+fs_rw_hugetlbfs_files(httpd_t)
 
 auth_use_nsswitch(httpd_t)
 
diff --git a/policy/modules/services/glusterd.fc b/policy/modules/services/glusterd.fc
index 735bd28..9399553 100644
--- a/policy/modules/services/glusterd.fc
+++ b/policy/modules/services/glusterd.fc
@@ -6,6 +6,12 @@
 /usr/sbin/glusterd	--	gen_context(system_u:object_r:glusterd_initrc_exec_t,s0)
 /usr/sbin/glusterfsd	--	gen_context(system_u:object_r:glusterd_exec_t,s0)
 
+/usr/sbin/glustereventsd   --  gen_context(system_u:object_r:glusterd_exec_t,s0)
+/usr/sbin/gluster-eventsapi   --       gen_context(system_u:object_r:glusterd_exec_t,s0)
+
+/usr/libexec/glusterfs/peer_eventsapi.py    --         gen_context(system_u:object_r:glusterd_exec_t,s0)
+/usr/libexec/glusterfs/events/glustereventsd.py   --   gen_context(system_u:object_r:glusterd_exec_t,s0)
+
 /opt/glusterfs/[^/]+/sbin/glusterfsd	--	gen_context(system_u:object_r:glusterd_exec_t,s0)
 
 /var/lib/glusterd(/.*)?		gen_context(system_u:object_r:glusterd_var_lib_t,s0)
diff --git a/policy/modules/services/glusterd.te b/policy/modules/services/glusterd.te
index 4aa6cac..4b815d6 100644
--- a/policy/modules/services/glusterd.te
+++ b/policy/modules/services/glusterd.te
@@ -60,7 +60,7 @@ files_type(glusterd_brick_t)
 # Local policy
 #
 
-allow glusterd_t self:capability { sys_admin sys_resource sys_ptrace dac_override chown dac_read_search fowner fsetid kill setgid setuid net_admin mknod net_raw };
+allow glusterd_t self:capability { sys_admin sys_resource sys_ptrace dac_override chown dac_read_search fowner fsetid kill setgid setuid net_admin mknod net_raw ipc_lock };
 
 allow glusterd_t self:process { getcap setcap setrlimit signal_perms setsched getsched };
 allow glusterd_t self:sem create_sem_perms;
@@ -98,6 +98,7 @@ relabel_files_pattern(glusterd_t, glusterd_var_lib_t, glusterd_var_lib_t)
 
 manage_dirs_pattern(glusterd_t, glusterd_brick_t, glusterd_brick_t)
 manage_files_pattern(glusterd_t, glusterd_brick_t, glusterd_brick_t)
+manage_sock_files_pattern(glusterd_t, glusterd_brick_t, glusterd_brick_t)
 manage_fifo_files_pattern(glusterd_t, glusterd_brick_t, glusterd_brick_t)
 manage_lnk_files_pattern(glusterd_t, glusterd_brick_t, glusterd_brick_t)
 manage_blk_files_pattern(glusterd_t, glusterd_brick_t, glusterd_brick_t)
@@ -195,6 +196,7 @@ userdom_read_user_tmp_files(glusterd_t)
 userdom_delete_user_tmpfs_files(glusterd_t)
 userdom_rw_user_tmpfs_files(glusterd_t)
 userdom_kill_all_users(glusterd_t)
+userdom_signal_unpriv_users(glusterd_t)
 
 mount_domtrans(glusterd_t)
 
